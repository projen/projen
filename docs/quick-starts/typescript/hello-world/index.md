---
sidebar_position: 1
---

# Hello World with TypeScript

To create a new TypeScript project, use `npx projen new typescript`:

```shell
npx projen new typescript
```

This will create a `.projenrc.ts` file which defines your project.

`npx projen new` will synthesize a standard project directory structure with some sample
code.

```shell
├── src
│   ├── index.ts
├── test
│   ├── hello.test.ts
├── node_modules
├── package.json
└── tsconfig.json
```

The default options will set up your `tsconfig` root directory as `src` and your
out directory as `lib`. By default the `test` directory is not compiled to
JavaScript and instead compiled when `npx projen test` is executed.

The `npx projen new` command will also generate a `.projenrc.js` file which includes
the definition of your project with any options you specified in the command
line:

```js
const { typescript } = require("projen");
const project = new typescript.TypeScriptProject({
  defaultReleaseBranch: "main",
  name: "my-project",
});
project.synth();
```

To modify your project definitions, edit `.projenrc.ts` and run `projen` again
to re-synthesize your project. The following sections describe the various
features of your TypeScript project.

## Dependencies

You can specify dependencies in your project via the `deps`, `devDeps`, and
`peerDeps` options similar to what you might expect in a `package.json` file.

The recommendation is to only specify the module name here (e.g. `express`).
This will behave similar to yarn add or npm install in the sense that it will
add the module as a dependency to your package.json file with the latest version
(`^`). You can specify semver requirements in the same syntax passed to npm i or
yarn add (e.g. `express@^2`) and this will be what your `package.json` will
eventually include.

## Migrating your TypeScript Project to Projen

Projen provides some opinionated defaults that may be different than the defaults
you rely on in your existing TypeScript Project. This section attempts to document
some of the common themes to look out for when migrating an existing TypeScript
project to Projen.

## Migrating .projenrc.js to .projenrc.ts

If you'd like to upgrade an existing project that was creating using a JS-based
projen RC file (.projenrc.js) to a TS-based projen RC file (.projenrc.ts):

1. Add `projenrcTs: true` to your project.
2. Run `npx projen`.
3. Rename `.projenrc.js` to `.projenrc.ts`.
4. Update `require`s to `import`s.
5. Run `npx projen`.

### Default Directory Structure

The most important is the default project structure. Projen expects all source code
to be in the `src` directory and will write compiled `.js` and `.d.ts` files to the
`lib` directory. Test files are expected in the `test` directory and not compiled
to javascript. Instead, Projen configures `Jest` to compile `.ts` files when testing.

As this is only one of many structures for valid TypeScript projects, your migrated
project may not work out-of-the-box with Projen. We recommend that you take the
time to fully migrate to the expected project structure so that you can take full
advantage of Projen management on your repository.

However, you can also configure your `.projenrc.js` file to reflect your current
project structure. A `TypeScriptProject` exposes all options in a `tsconfig` file
so you can manually set the structure you want. Here is an example for specifying a
structure where the TypeScript compiler finds all TypeScript files recursively in your
repository:

```js
const { typescript } = require('projen');
const project = new typescript.TypeScriptProject({
  ...
  tsconfig: {
    compilerOptions: {
      rootDir: '.',
      outDir: '.',
    },
    include: ['**/*.ts'], // all typescript files recursively
  },
});
```

### Dependency Versions

As described in the section above, Projen recommends that you list your dependencies
only by module name and have Projen install the latest version of the package. A
consequence of this recommendation is that when migrating, you may unknowingly update
your dependencies to incompatible versions. You can always provide specific semver
requirements (e.g. `express@2.1.0`) if necessary.
