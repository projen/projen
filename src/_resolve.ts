import { types } from "util";
import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // Special resolution for few JavaScript built-in types
  // that by default would be stringified as empty objects ('{}')
  // as they are missing a `toJSON` implementation.
  switch (true) {
    case types.isRegExp(value):
      if (value.flags) {
        throw new Error(
          "RegExp with flags should be explicitly converted to a string"
        );
      }
      return value.source;

    case types.isSet(value):
      return resolve(Array.from(value), options);

    case types.isMap(value):
      return resolve(Object.fromEntries(value), options);

    case types.isBoxedPrimitive(value):
      return resolve(value.valueOf(), options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
