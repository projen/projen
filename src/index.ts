export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
