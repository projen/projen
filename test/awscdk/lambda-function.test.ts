import { mkdirSync, writeFileSync } from "fs";
import { join } from "path";
import { DependencyType } from "../../src";
import * as awscdk from "../../src/awscdk";
import { AwsCdkDepsJs } from "../../src/awscdk/awscdk-deps-js";
import { Testing } from "../../src/testing";
import { TypeScriptProject } from "../../src/typescript";

describe("bundled function", () => {
  let generatedSource: string;
  let tasks: Record<string, any>;
  let npmignore: string[];
  let gitignore: string[];

  beforeEach(() => {
    const project = new TypeScriptProject({
      name: "hello",
      defaultReleaseBranch: "main",
      bundlerOptions: {
        assetsDir: "my-assets",
      },
    });

    new awscdk.LambdaFunction(project, {
      entrypoint: join("src", "hello.lambda.ts"),
      cdkDeps: cdkDepsForProject(project),
    });

    const snapshot = Testing.synth(project);

    generatedSource = snapshot["src/hello-function.ts"];
    tasks = snapshot[".projen/tasks.json"].tasks;
    npmignore = snapshot[".npmignore"].split("\n");
    gitignore = snapshot[".gitignore"].split("\n");
  });

  test("generates source code for a lambda construct", () => {
    expect(generatedSource).toMatchSnapshot();
  });

  test("creates a single project-wide bundle task", () => {
    expect(tasks.bundle).toEqual({
      description: "Prepare assets",
      name: "bundle",
      steps: [{ spawn: "bundle:hello.lambda" }],
    });
  });

  test("creates a specific bundle task for this function", () => {
    expect(tasks["bundle:hello.lambda"]).toEqual({
      description: "Create a JavaScript bundle from src/hello.lambda.ts",
      name: "bundle:hello.lambda",
      steps: [
        {
          exec: 'esbuild --bundle src/hello.lambda.ts --target="node22" --platform="node" --outfile="my-assets/hello.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:@aws-sdk/*',
        },
      ],
    });
  });

  test("spawns the bundle task as part of pre-compilation", () => {
    expect(tasks["pre-compile"].steps).toStrictEqual([{ spawn: "bundle" }]);
  });

  test("includes the bundle directory inside the node package but not commit to source control", () => {
    expect(npmignore).toContain("!/my-assets/");
    expect(gitignore).toContain("/my-assets/");
  });
});

test("fails if entrypoint does not have the .lambda suffix", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });
  expect(
    () =>
      new awscdk.LambdaFunction(project, {
        entrypoint: join("src", "hello-no-lambda.ts"),
        cdkDeps: cdkDepsForProject(project),
      })
  ).toThrow(
    "hello-no-lambda.ts must have a .lambda.ts or .edge-lambda.ts extension"
  );
});

test("constructFile and constructName can be used to customize the generated construct", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.lambda.ts"),
    constructFile: "my-construct.ts",
    constructName: "MyConstruct",
    cdkDeps: cdkDepsForProject(project),
  });

  const snapshot = Testing.synth(project);
  const generatedSource = snapshot["src/my-construct.ts"];
  expect(generatedSource).toMatchSnapshot();
});

test("runtime can be used to customize the lambda runtime Node 14.x and esbuild target", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.lambda.ts"),
    runtime: awscdk.LambdaRuntime.NODEJS_14_X,
    cdkDeps: cdkDepsForProject(project),
  });

  const snapshot = Testing.synth(project);
  const generatedSource = snapshot["src/hello-function.ts"];
  const tasks = snapshot[".projen/tasks.json"].tasks;
  expect(generatedSource).toContain(
    "runtime: new lambda.Runtime('nodejs14.x', lambda.RuntimeFamily.NODEJS),"
  );
  expect(tasks["bundle:hello.lambda"]).toEqual({
    description: "Create a JavaScript bundle from src/hello.lambda.ts",
    name: "bundle:hello.lambda",
    steps: [
      {
        exec: 'esbuild --bundle src/hello.lambda.ts --target="node14" --platform="node" --outfile="assets/hello.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:aws-sdk',
      },
    ],
  });
});

test("runtime can be used to customize the lambda runtime Node 16.x and esbuild target", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.lambda.ts"),
    runtime: awscdk.LambdaRuntime.NODEJS_16_X,
    cdkDeps: cdkDepsForProject(project),
  });

  const snapshot = Testing.synth(project);
  const generatedSource = snapshot["src/hello-function.ts"];
  const tasks = snapshot[".projen/tasks.json"].tasks;
  expect(generatedSource).toContain(
    "runtime: new lambda.Runtime('nodejs16.x', lambda.RuntimeFamily.NODEJS),"
  );
  expect(tasks["bundle:hello.lambda"]).toEqual({
    description: "Create a JavaScript bundle from src/hello.lambda.ts",
    name: "bundle:hello.lambda",
    steps: [
      {
        exec: 'esbuild --bundle src/hello.lambda.ts --target="node16" --platform="node" --outfile="assets/hello.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:aws-sdk',
      },
    ],
  });
});

test.each([
  awscdk.LambdaRuntime.NODEJS_18_X,
  awscdk.LambdaRuntime.NODEJS_20_X,
  awscdk.LambdaRuntime.NODEJS_22_X,
])(
  "runtime can be used to customize the lambda runtime $functionRuntime and esbuild target",
  (runtime) => {
    const project = new TypeScriptProject({
      name: "hello",
      defaultReleaseBranch: "main",
    });

    new awscdk.LambdaFunction(project, {
      entrypoint: join("src", "hello.lambda.ts"),
      runtime,
      cdkDeps: cdkDepsForProject(project),
    });

    const snapshot = Testing.synth(project);
    const generatedSource = snapshot["src/hello-function.ts"];
    const tasks = snapshot[".projen/tasks.json"].tasks;
    expect(generatedSource).toContain(
      `runtime: new lambda.Runtime('${runtime.functionRuntime}', lambda.RuntimeFamily.NODEJS),`
    );
    expect(tasks["bundle:hello.lambda"]).toEqual({
      description: "Create a JavaScript bundle from src/hello.lambda.ts",
      name: "bundle:hello.lambda",
      steps: [
        {
          exec: `esbuild --bundle src/hello.lambda.ts --target="${runtime.esbuildTarget}" --platform="node" --outfile="assets/hello.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:@aws-sdk/*`,
        },
      ],
    });
  }
);

test("aws sdk v3 packages are considered external with NODEJS_18_X", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.lambda.ts"),
    runtime: awscdk.LambdaRuntime.NODEJS_18_X,
    cdkDeps: cdkDepsForProject(project),
  });

  const snapshot = Testing.synth(project);
  const tasks = snapshot[".projen/tasks.json"].tasks;
  expect(tasks["bundle:hello.lambda"]).toEqual({
    description: "Create a JavaScript bundle from src/hello.lambda.ts",
    name: "bundle:hello.lambda",
    steps: [
      {
        exec: 'esbuild --bundle src/hello.lambda.ts --target="node18" --platform="node" --outfile="assets/hello.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:@aws-sdk/*',
      },
    ],
  });
});

test("AWS SDK connection reuse", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.lambda.ts"),
    cdkDeps: cdkDepsForProject(project),
  });

  const snapshot = Testing.synth(project);
  const generatedSource = snapshot["src/hello-function.ts"];
  expect(generatedSource).toContain(
    "this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });"
  );
});

test("AWS SDK connection reuse can be disabled", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.lambda.ts"),
    cdkDeps: cdkDepsForProject(project),
    awsSdkConnectionReuse: false,
  });

  const snapshot = Testing.synth(project);
  const generatedSource = snapshot["src/hello-function.ts"];
  expect(generatedSource).not.toContain(
    "this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });"
  );
});

test("Edge function", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.edge-lambda.ts"),
    cdkDeps: cdkDepsForProject(project),
    edgeLambda: true,
  });

  const snapshot = Testing.synth(project);
  const generatedSource = snapshot["src/hello-function.ts"];
  expect(generatedSource).toMatchSnapshot();
});

test("eslint allows handlers to import dev dependencies", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });
  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.lambda.ts"),
    cdkDeps: cdkDepsForProject(project),
  });
  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "world.lambda.ts"),
    cdkDeps: cdkDepsForProject(project),
  });

  const snapshot = Testing.synth(project);
  expect(
    snapshot[".eslintrc.json"].rules["import/no-extraneous-dependencies"]
  ).toStrictEqual([
    "error",
    {
      devDependencies: [
        "**/test/**",
        "**/build-tools/**",
        "src/hello.lambda.ts",
        "src/world.lambda.ts",
      ],
      optionalDependencies: false,
      peerDependencies: true,
    },
  ]);
});

test("esbuild dependency is added", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });
  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.lambda.ts"),
    cdkDeps: cdkDepsForProject(project),
  });
  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "world.lambda.ts"),
    cdkDeps: cdkDepsForProject(project),
  });

  const snapshot = Testing.synth(project);
  const deps = snapshot[".projen/deps.json"].dependencies;
  expect(deps.filter((d: any) => d.name === "esbuild").length).toEqual(1);
});

test("multiple functions", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });
  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.lambda.ts"),
    cdkDeps: cdkDepsForProject(project),
  });
  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "world.lambda.ts"),
    cdkDeps: cdkDepsForProject(project),
  });

  const snapshot = Testing.synth(project);
  expect(snapshot["src/hello-function.ts"]).toMatchSnapshot();
  expect(snapshot["src/world-function.ts"]).toMatchSnapshot();
});

test("auto-discover", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  // create fake handlers
  const srcdir = join(project.outdir, project.srcdir);
  mkdirSync(join(srcdir, "subdir"), { recursive: true });
  writeFileSync(
    join(srcdir, "hello.lambda.ts"),
    "export function handler() {}"
  );
  writeFileSync(
    join(srcdir, "subdir", "world.lambda.ts"),
    "export function handler() {}"
  );
  writeFileSync(
    join(srcdir, "subdir", "jangy.lambda.ts"),
    "export function handler() {}"
  );

  new awscdk.AutoDiscover(project, {
    srcdir: project.srcdir,
    testdir: project.testdir,
    tsconfigPath: project.tsconfigDev.fileName,
    cdkDeps: new AwsCdkDepsJs(project, {
      cdkVersion: "1.23.0",
      dependencyType: DependencyType.RUNTIME,
    }),
    lambdaOptions: {
      runtime: awscdk.LambdaRuntime.NODEJS_22_X,
    },
  });

  const snapshot = Testing.synth(project);
  expect(snapshot["src/hello-function.ts"]).toMatchSnapshot();
  expect(snapshot["src/subdir/world-function.ts"]).toMatchSnapshot();
  expect(snapshot["src/subdir/jangy-function.ts"]).toMatchSnapshot();
  const tasks = snapshot[".projen/tasks.json"].tasks;
  expect(tasks.bundle).toMatchSnapshot();

  const expected = [
    "bundle:hello.lambda",
    "bundle:hello.lambda:watch",
    "bundle:subdir/jangy.lambda",
    "bundle:subdir/jangy.lambda:watch",
    "bundle:subdir/world.lambda",
    "bundle:subdir/world.lambda:watch",
  ];

  for (const name of expected) {
    expect(tasks[name]).not.toBeUndefined();
    expect(tasks[name]).toMatchSnapshot();
  }
});

test("generates cdkv2-compatible imports", () => {
  const project = new TypeScriptProject({
    name: "hello",
    defaultReleaseBranch: "main",
  });

  new awscdk.LambdaFunction(project, {
    entrypoint: join("src", "hello.lambda.ts"),
    cdkDeps: cdkDepsForProject(project, "2.3.1"),
  });

  const snapshot = Testing.synth(project);
  expect(snapshot["src/hello-function.ts"]).toMatchSnapshot();
});

function cdkDepsForProject(
  project: TypeScriptProject,
  cdkVersion = "1.0.0"
): awscdk.AwsCdkDeps {
  return new AwsCdkDepsJs(project, {
    cdkVersion: cdkVersion,
    dependencyType: DependencyType.RUNTIME,
  });
}
