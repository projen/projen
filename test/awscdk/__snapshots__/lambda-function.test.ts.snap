// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edge function 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import * as path from 'path';
import * as cloudfront from '@aws-cdk/aws-cloudfront';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for HelloFunction
 */
export interface HelloFunctionProps extends cloudfront.experimental.EdgeFunctionProps {
}

/**
 * An AWS Lambda function which executes src/hello.
 */
export class HelloFunction extends cloudfront.experimental.EdgeFunction {
  constructor(scope: Construct, id: string, props?: HelloFunctionProps) {
    super(scope, id, {
      description: 'src/hello.edge-lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs22.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/hello.edge-lambda')),
    });
  }
}"
`;

exports[`auto-discover 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for HelloFunction
 */
export interface HelloFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/hello.
 */
export class HelloFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: HelloFunctionProps) {
    super(scope, id, {
      description: 'src/hello.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs22.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/hello.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}"
`;

exports[`auto-discover 2`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for WorldFunction
 */
export interface WorldFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/subdir/world.
 */
export class WorldFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: WorldFunctionProps) {
    super(scope, id, {
      description: 'src/subdir/world.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs22.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../../assets/subdir/world.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}"
`;

exports[`auto-discover 3`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for JangyFunction
 */
export interface JangyFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/subdir/jangy.
 */
export class JangyFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: JangyFunctionProps) {
    super(scope, id, {
      description: 'src/subdir/jangy.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs22.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../../assets/subdir/jangy.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}"
`;

exports[`auto-discover 4`] = `
{
  "description": "Prepare assets",
  "name": "bundle",
  "steps": [
    {
      "spawn": "bundle:hello.lambda",
    },
    {
      "spawn": "bundle:subdir/jangy.lambda",
    },
    {
      "spawn": "bundle:subdir/world.lambda",
    },
  ],
}
`;

exports[`auto-discover 5`] = `
{
  "description": "Create a JavaScript bundle from src/hello.lambda.ts",
  "name": "bundle:hello.lambda",
  "steps": [
    {
      "exec": "esbuild --bundle src/hello.lambda.ts --target="node22" --platform="node" --outfile="assets/hello.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:@aws-sdk/*",
    },
  ],
}
`;

exports[`auto-discover 6`] = `
{
  "description": "Continuously update the JavaScript bundle from src/hello.lambda.ts",
  "name": "bundle:hello.lambda:watch",
  "steps": [
    {
      "exec": "esbuild --bundle src/hello.lambda.ts --target="node22" --platform="node" --outfile="assets/hello.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:@aws-sdk/* --watch",
    },
  ],
}
`;

exports[`auto-discover 7`] = `
{
  "description": "Create a JavaScript bundle from src/subdir/jangy.lambda.ts",
  "name": "bundle:subdir/jangy.lambda",
  "steps": [
    {
      "exec": "esbuild --bundle src/subdir/jangy.lambda.ts --target="node22" --platform="node" --outfile="assets/subdir/jangy.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:@aws-sdk/*",
    },
  ],
}
`;

exports[`auto-discover 8`] = `
{
  "description": "Continuously update the JavaScript bundle from src/subdir/jangy.lambda.ts",
  "name": "bundle:subdir/jangy.lambda:watch",
  "steps": [
    {
      "exec": "esbuild --bundle src/subdir/jangy.lambda.ts --target="node22" --platform="node" --outfile="assets/subdir/jangy.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:@aws-sdk/* --watch",
    },
  ],
}
`;

exports[`auto-discover 9`] = `
{
  "description": "Create a JavaScript bundle from src/subdir/world.lambda.ts",
  "name": "bundle:subdir/world.lambda",
  "steps": [
    {
      "exec": "esbuild --bundle src/subdir/world.lambda.ts --target="node22" --platform="node" --outfile="assets/subdir/world.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:@aws-sdk/*",
    },
  ],
}
`;

exports[`auto-discover 10`] = `
{
  "description": "Continuously update the JavaScript bundle from src/subdir/world.lambda.ts",
  "name": "bundle:subdir/world.lambda:watch",
  "steps": [
    {
      "exec": "esbuild --bundle src/subdir/world.lambda.ts --target="node22" --platform="node" --outfile="assets/subdir/world.lambda/index.js" --tsconfig="tsconfig.dev.json" --external:@aws-sdk/* --watch",
    },
  ],
}
`;

exports[`bundled function generates source code for a lambda construct 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for HelloFunction
 */
export interface HelloFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/hello.
 */
export class HelloFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: HelloFunctionProps) {
    super(scope, id, {
      description: 'src/hello.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs22.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../my-assets/hello.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}"
`;

exports[`constructFile and constructName can be used to customize the generated construct 1`] = `undefined`;

exports[`generates cdkv2-compatible imports 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import * as path from 'path';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { Construct } from 'constructs';

/**
 * Props for HelloFunction
 */
export interface HelloFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/hello.
 */
export class HelloFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: HelloFunctionProps) {
    super(scope, id, {
      description: 'src/hello.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs22.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/hello.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}"
`;

exports[`multiple functions 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for HelloFunction
 */
export interface HelloFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/hello.
 */
export class HelloFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: HelloFunctionProps) {
    super(scope, id, {
      description: 'src/hello.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs22.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/hello.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}"
`;

exports[`multiple functions 2`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import * as path from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct } from '@aws-cdk/core';

/**
 * Props for WorldFunction
 */
export interface WorldFunctionProps extends lambda.FunctionOptions {
}

/**
 * An AWS Lambda function which executes src/world.
 */
export class WorldFunction extends lambda.Function {
  constructor(scope: Construct, id: string, props?: WorldFunctionProps) {
    super(scope, id, {
      description: 'src/world.lambda.ts',
      ...props,
      runtime: new lambda.Runtime('nodejs22.x', lambda.RuntimeFamily.NODEJS),
      handler: 'index.handler',
      code: lambda.Code.fromAsset(path.join(__dirname, '../assets/world.lambda')),
    });
    this.addEnvironment('AWS_NODEJS_CONNECTION_REUSE_ENABLED', '1', { removeInEdge: true });
  }
}"
`;
